File: whiteboard-server/src/whiteboard/event-emitters/event-emitter.ts
export type EventListener = {
  handleEvent: (event: string, data: any) => void;
};

export class EventEmitter {
  private eventListeners = new Map<string, EventListener[]>();

  on = (event: string, listener: EventListener) => {
    const listeners = this.eventListeners.get(event);

    if (listeners) {
      listeners.push(listener);
    } else {
      this.eventListeners.set(event, [listener]);
    }
  };

  emit = (event: string, data: unknown) => {
    const listeners = this.eventListeners.get(event);

    if (listeners) {
      for (const listener of listeners) {
        listener.handleEvent(event, data);
      }
    }
  };
}

File: whiteboard-app/src/main/java/dev/oscarrojas/whiteboard/EventEmitter.java
package dev.oscarrojas.whiteboard;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Component;

@Component
public class EventEmitter {

  private Map<String, List<EventListener>> listeners;

  EventEmitter(List<EventListener> listeners) {
    this.listeners = new HashMap<>();

    for (EventListener listener : listeners) {
      addEventListener(listener);
    }

  }

  public void emit(String event, AppSession session, byte[] payload) {
    List<EventListener> eventListeners = listeners.get(event);

    if (eventListeners == null) {
      return;
    }

    for (EventListener listener : eventListeners) {
      listener.handleEvent(event, session, payload);
    }
  }

  public void addEventListener(EventListener listener) {

    for (String event : listener.getEvents()) {

      List<EventListener> eventListeners = listeners.get(event);

      if (eventListeners == null) {
        listeners.put(event, new ArrayList<>(Arrays.asList(listener)));
      } else {
        eventListeners.add(listener);
      }

    }

  }

}
